fix: Upgraded Mathjax version from 2.7 to 3.0.1

Yagnesh1998 opened the pull request:
**Description:**
fix: Upgraded Mathjax version from 2.7 to 3.0.1

OpenEdx is using Mathjax 2.7 version in the current version which is less accessible this drives us to upgrade it to the new latest version of Mathjax (3.0.1) which is having more support for accessible people.
Example:
Old (2.7): when we tab to the equation, it doesn't speak.
New (3.0.1) when we tab to the equation, it starts to speak the equation nicely.

**Testing instructions:**

- Open any course content page where we have an equation on content.
- Enable accessible mode for the application.
- Start tabbing and reach the equation, audio is not played for the equation.
- After the upgrade, it starts to play equations nicely.

Commit:
fix: Upgraded Mathjax version from 2.7 to 3.0.1

openedx-webhooks commented:
Thanks for the pull request, @Yagnesh1998! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

e0d committed:
Merge branch 'master' into edx-mathjax

e0d commented:
Updated with a merge commit.

e0d committed:
Merge branch 'master' into edx-mathjax

e0d commented:
@Yagnesh1998 looks like there are test failures, can you have a look.

Yagnesh1998 commented:
@e0d can u please help what is failure?

e0d commented:
You will need to look at the log of the failing unit test and figure out why the change is causing it.  You can see the logs by clicking on the details link.  I'm going to update the branch, but the failures are almost certainly caused by the upgrade.

e0d committed:
Merge branch 'master' into edx-mathjax

Yagnesh1998 committed:
Merge branch 'master' into edx-mathjax

e0d committed:
Merge branch 'master' into edx-mathjax

e0d commented:
@Yagnesh1998 The failures seems to be related to differences in MathJax configuration.  See this:

https://github.com/openedx/edx-platform/actions/runs/5501574151/jobs/10025256535?pr=32418#step:7:100

e0d committed:
Merge branch 'master' into edx-mathjax

arbrandes commented:
This is on my queue to review, but one thing off the bat: MathJax upgrades must not be taken lightly.  There is a lot of content across organizations that uses it, and the relevant stakeholders have to be brought into this so they can check for regressions.  (There have been issues with MathJax upgrades in the past.)

With that in mind, ideally we'd find a way for content owners to roll this out incrementally.  Maybe via course waffle flag to enable MathJax 3?

mphilbrick211 commented:
> @Yagnesh1998 The failures seems to be related to differences in MathJax configuration. See this:
> 
> https://github.com/openedx/edx-platform/actions/runs/5501574151/jobs/10025256535?pr=32418#step:7:100

Hi @Yagnesh1998! Just following up on this :)

wittjeff commented:
The current supported version of MathJax is 3.2.2. https://www.mathjax.org/

As noted above, the configuration is different between v2.x and v3.x.  The platform currently uses a mix of v.2.x and v3.x (v3.x in Discussions, for example).  

There is a known shortcoming with v.3.x related to lack of support for line breaks (or automatic wrapping, or both).  https://github.com/mathjax/MathJax/issues/2312 
https://github.com/fred-wang/webextension-native-mathml/issues/14
This issue is apparently resolved in v4.0a. https://github.com/mathjax/MathJax-src/releases/tag/4.0.0-alpha.1 

The interactivity (particularly accessibility features) of v3 is quite a bit different than v3, so users who rely on accessibility features may consider an automatic upgrade disruptive.  More specifically, institutions that host courses that include MathML, and who provide accessibility documentation of their own, may find that their docs are obsolete.

In the last year or so, Firefox and Chrome have added support for native rendering of MathML (though I'm not sure about details re expected formats).  If the browser supports minimal rendering of MathML, it's conceivable that different users having more specific accessibility UI preferences might be able to use their preferred solutions through browser extensions instead of integration in the platform.  A long-term solution that doesn't involve pushing JS packages for MathML rendering might also save a lot of $ in data traffic.

There's another accessibility solution for MathML, called MathCAT. https://nsoiffer.github.io/MathCAT/ 
The maintainer is well regarded and the feature list is impressive but I'm not deeply familiar with it and I don't know anyone who uses it.

What I'd like to see in Open edX learning experience MFE and supported xBlocks: 
Add a course-level setting (in Studio) for MathML rendering, with 5 or 6 options:
- None (rely on browser rendering, if confirmed to be adequate)
- MathJax v2.x (last supported version) -- default value
- MathJax v3.x  (last supported version)
- MathJax v4.0ax (last preview version)
- MathCAT (last supported version)
- possible 6th option: KaTeX (last supported version) https://katex.org/

This way, course authors can confirm layout compatibility with their MathML content (something we can't do in an automated way), and can also also toggle accessibility solutions if a learner has a specific preference.

For the rest of the platform, I think upgrading to MathJax v3.x (last supported) is safe for now.

wittjeff commented:
Possibly of interest: Force MathJax/KaTeX/MediaWiki to use native MathML rendering. https://chrome.google.com/webstore/detail/native-mathml/lcadkfljmcmcflpdbfmgcpjlejmpcplg 

Related to addition of MathML rendering support in Chrome:
https://famousandfaded.com/deep-cuts/mathml-chrome-cross-platform-support/

I see that for native rendering, Chrome is looking for <math> syntax, so there would be some updating needed in Open edX docs, and also some compatibility checking and/or adjustment with math authoring solutions used in TinyMCE throughout the platform.

Open edX expects authors to use
\( equation \)
or
[mathjaxinline] equation [/mathjaxinline]

https://edx.readthedocs.io/projects/open-edx-building-and-running-a-course/en/latest/exercises_tools/mathjax.html

MathType authoring plugin for TinyMCE is another possible accessible authoring solution:  https://docs.wiris.com/mathtype/en/mathtype-integrations/mathtype-for-html-editors/mathtype-for-tinymce.html

https://www.npmjs.com/package/@wiris/mathtype-tinymce6

Not mentioning edX though: https://www.wiris.com/en/mathtype/mathtype-for-lms/

wittjeff commented:
Sample pages for testing rendering: https://developer.mozilla.org/en-US/docs/Web/MathML/Examples

wittjeff commented:
One more thought: I'm told that MathJax isn't working in some countries such as China, due to certain content servers being blocked. If we're going to update configuration, it might be a good time to move the source files to a self-managed server to resolve this issue.

mphilbrick211 commented:
Hi @Yagnesh1998! Flagging the above comments for you. Please let us know if you plan to pursue this pull request. Thanks!

arbrandes commented:
Seconding @wittjeff's comment above: I don't think we'll be able to straight-up upgrade Mathjax to v3.  At the very least we'll need to retain the current (v2) behavior as a default during a deprecation window, while giving the option to turn on the new one.  And in an ideal world we'd give authors the option to pick, as Jeff describes.

Yagnesh1998 commented:
**I am very sorry for the very late replay**  @mphilbrick211 & @wittjeff

Yagnesh1998 commented:
@wittjeff & @arbrandes 
I ran into a problem when I made a pull request this time.
 Blind people can't see they are only here and they also access the edx-platform. we have installed a screen reader extension in Chrome and we have added some Mathjax equations but there is no support so we have made some changes on the platform side and Version upgraded.  We ran this extension again and that time it easily worked all Mathjax functions correctly. we added those Mathjax functions in Studio and checked in LMS on 12/06/2023 so this time we thought this issue of the Mathjax version so I made a pull request. please let me know what I need to do next.
 now I can see my commit is conflicts.

mphilbrick211 commented:
Hi @wittjeff and @arbrandes - just flagging this for follow-up.

navinkarkera commented:
I have created a PR: https://github.com/openedx/edx-platform/pull/33555 for the full upgrade. Apologies for missing this PR but it seems to be incomplete, still I should have probably continued the work here instead of starting fresh.
cc: @Yagnesh1998 @arbrandes

openedx-webhooks commented:
@Yagnesh1998 Even though your pull request wasn’t merged, please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/32418&entry.1484655318=https://github.com/Yagnesh1998&entry.752974735=2023-06-12+06:15&entry.1917517419=2023-11-08+08:10&entry.2133058324=No) so we can improve your experience in the future.

<!-- comment:end_survey -->

Yagnesh1998 commented:
@navinkarkera I close this PR and we will continue on [#33555](https://github.com/openedx/edx-platform/pull/33555)

wittjeff commented:
More notes just to be thorough:
- @ormsbee is disinclined to pull MathJax libraries into our codebase and I'm sympathetic to his concerns.
- If MathJax (or any other option from the suggested options above) doesn't load as expected, we should default to some other option.
- I note just as a curiosity that as of this date ChatGPT4.0 uses KaTeX to render math output.  I don't know if that will be inherited in any of edX.org's chatbot implementations when we eventually upgrade (v3.5T apparently "doesn't do math").
